provider/uprops/src/enumerated.rs

```diff
@@ -0,0 +1,306 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use crate::error::Error;
+use crate::upropdump_serde;
+use icu_provider::iter::IterableDataProviderCore;
+use icu_provider::prelude::*;
+use icu_uniset::provider::*;
+use icu_uniset::UnicodeSetBuilder;
+use std::borrow::Cow;
+use std::fs;
+use std::path::PathBuf;
+
+pub struct EnumeratedPropertiesDataProvider {
+    root_dir: PathBuf,
+}
+
+/// A data provider reading from .toml files produced by the ICU4C upropdump tool.
+impl EnumeratedPropertiesDataProvider {
+    pub fn new(root_dir: PathBuf) -> Self {
+        EnumeratedPropertiesDataProvider { root_dir }
+    }
+    fn get_toml_data(&self, name: &str) -> Result<upropdump_serde::enumerated::Main, Error> {
+        let mut path: PathBuf = self.root_dir.clone().join(name);
+        path.set_extension("toml");
+        let toml_str = fs::read_to_string(&path).map_err(|e| Error::Io(e, path.clone()))?;
+        toml::from_str(&toml_str).map_err(|e| Error::Toml(e, path))
+    }
+}
+
+fn expand_groupings<'a>(prop_name: &str, prop_val: &'a str) -> Vec<&'a str> {
```

Comment:
Question: Do we want to build different data keys for `gc=L`, `gc=LC`, and the other expansions, or do we want to only have data keys for the "leaf nodes", and combine them at runtime?  What are the pros and cons?

Reply:
I expect that `gc=L` is going to be used much more often than, say, `gc=Lu`, so it's awkward to make the former case slower than the latter. This will be particularly relevant once #855 makes UnicodeSet zero-copy in the not-combined-at-runtime case: it would be a shame if we only get the zero-copy performance in the uncommon case.

The downside, of course, is that we store extra data, although it's less than twice as much because (for example) a run of N uppercase/lowercase pairs is 2N ranges in `gc=Lu` and `gc=Ll`, but a single range in `gc=L`. Based on a quick test, this effect cuts the number of ranges roughly in half, so we're looking at a ~50% increase in data size.

In a world where we're storing property data as zero-copy inversion lists, I think separate keys makes sense. As we add support for codepoint tries, it's possible that we want to unify the backing data and generate the inversion list on the fly, but in that case we're likely moving to a single "gc" key instead of "gc=*", so the question is moot.

Reply:
Yes, I agree with your reasoning. Thanks for explaining.